Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/diagnosiskeyprovider/DefaultDiagnosisKeyProvider.kt

```diff
@@ -22,95 +19,37 @@ class DefaultDiagnosisKeyProvider @Inject constructor(
     private val enfClient: ExposureNotificationClient
 ) : DiagnosisKeyProvider {
 
-    override suspend fun provideDiagnosisKeys(
-        keyFiles: Collection<File>,
-        configuration: ExposureConfiguration?,
-        token: String
-    ): Boolean {
-        return try {
-            if (keyFiles.isEmpty()) {
-                Timber.d("No key files submitted, returning early.")
-                return true
-            }
-
-            val usedConfiguration = if (configuration == null) {
-                Timber.w("Passed configuration was NULL, creating fallback.")
-                ExposureConfiguration.ExposureConfigurationBuilder().build()
-            } else {
-                configuration
-            }
-
-            if (googleAPIVersion.isAtLeast(GoogleAPIVersion.V16)) {
-                provideKeys(keyFiles, usedConfiguration, token)
-            } else {
-                provideKeysLegacy(keyFiles, usedConfiguration, token)
-            }
-        } catch (e: Exception) {
-            Timber.e(
-                e, "Error during provideDiagnosisKeys(keyFiles=%s, configuration=%s, token=%s)",
-                keyFiles, configuration, token
-            )
-            throw e
+    override suspend fun provideDiagnosisKeys(keyFiles: Collection<File>): Boolean {
+        if (keyFiles.isEmpty()) {
+            Timber.d("No key files submitted, returning early.")
+            return true
         }
-    }
 
-    private suspend fun provideKeys(
-        files: Collection<File>,
-        configuration: ExposureConfiguration,
-        token: String
-    ): Boolean {
-        Timber.d("Using non-legacy key provision.")
-
-        if (!submissionQuota.consumeQuota(1)) {
-            Timber.w("Not enough quota available.")
-            // TODO Currently only logging, we'll be more strict in a future release
-            // return false
+        if (!googleAPIVersion.isAtLeast(GoogleAPIVersion.V15)) {
+            // Actually this shouldn't happen
+            Timber.d("No key files submitted because client uses an old unsupported version")
+            return false
```

Comment:
https://github.com/corona-warn-app/cwa-app-android/pull/1643/files#r525859782

Reply:
Short sum-up:

@d4rken proposed creating `googleAPIVersion.requireV15()` and throwing a specific exception at this point, propagated to the current error dialog, giving a specific error message like "Your systems exposure notification service are too old". The occurence of that exception is probably rare. Although the current error dialog is not nice UI/UX, could be enough and would be easy to implement.
  
@BMItter feels that's better than crashing, but user would be left alone with an exception here. It's proposed to check all the API/enf requirements while the app is starting.

My opinion:
Yes, I agree that's probably a rare condition. However, it's likely a nice initial starting point to enhance UX in error reporting and guiding to resolutions.
If we implement an exception/error dialog here (and `googleAPIVersion.requireV15()` had a similar structure like `googleAPIVersion.isAtLeast()`, just throwing an additional exception), there will be two different use cases when an exception is thrown/error dialog is presented: Either the ENF version is too low. Or the device is not supported (no ENF, API not connected (17)), see here:

https://github.com/corona-warn-app/cwa-app-android/blob/5bb2427e5671c681d6aad0d53c6a4772855abcb3/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/GoogleAPIVersion.kt#L21-L28

The two use cases will occur in these situations: A user reactivates an old phone for a family member for contact tracing. Or: A user reactivates an old phone to scan a COVID test QR code for a family member (e.g., child).
I'm not sure, how rare these use cases actually are.
If the OS of the reactivated phone basically supports ENF, it's just a matter of time until the new ENF version arrives at that phone. Until ENF update and with @d4rken 's proposed implementation, there will be an error dialog every single time CWA is checking for exposures in foreground. If the phone has no support for ENF, API 17 will be persistent, and again, with every single exposure check in foreground, CWA will present an error dialog. This can specifically be annoying, when you reactivated your phone only for retrieving a test result via QR code, and your're checking often whether it arrived.

So, my concluding proposal would be:
- Check for API not connected (Api 17) during onboarding or/and maybe also during `onCreate()` in MainActivity. Persist the result. Present a steady UI element to the user with further information (i.e., 'information (error) card'). This would go with @BMItter .
- Check for ENF version every time in `DefaultDiagnosisKeyProvider()` as the ENF version may update anytime. Persist fail and present information either in risk card for 'risk calculation stopped'-state (see #1615). or in the seperate 'information (error) card'.

What do you think?

Reply:
I think that would be an ideal solution, but time constraint do not allow for it in 1.8. This is too rare to warrant delaying the update. Throwing an exception that yields the error dialog is the way to go, IMHO. I'll discuss it with the team though.

@vaubaehn You raise valid points and we should explicitly discuss how the app should behave if the ENF version is not what we expect. Would you create a ticket for this?

The desired end result of the groundwork @BMItter has already provided with his new error tracking is in my opinion, a card that sits on the home screen that informs the user about "current issues that may impact the app", with options for help/details. The insufficient ENF version would be one such "issue" that could be displayed there. Unfortunately priorities did not yet align for this to get more attention, I'd like to link your idea and concerns to that.

Reply:
@d4rken 
After re-evaluating my own proposal and getting some fresh information by @MikeMcC399 [here ](https://covidapps.slack.com/archives/C0194ML0MLN/p1605814630477000?thread_ts=1605783709.474200&cid=C0194ML0MLN) , I also doubt that it's worth the effort for the time being.

> Throwing an exception that yields the error dialog is the way to go, IMHO.

Agreed.

> @vaubaehn You raise valid points and we should explicitly discuss how the app should behave if the ENF version is not what we expect. Would you create a ticket for this?

Yes, this is a better way to go. After re-evaluating my proposal, I stumbled across some wrong assumptions, e. g. `GoogleAPIVersion.isAtLeast()` does **not** throw an exception if API is not connected, so that issue will probably only come up once during onboarding (start tracing). Additionally, querying [ExposureNotificationStatus](https://developers.google.com/android/reference/com/google/android/gms/nearby/exposurenotification/ExposureNotificationStatus) in conjuction with API errors would provide more accurate information, and so on...
I'll try to create a ticket soon. 

> The desired end result of the groundwork @BMItter has already provided with his new error tracking is in my opinion, a card that sits on the home screen that informs the user about "current issues that may impact the app", with options for help/details. The insufficient ENF version would be one such "issue" that could be displayed there. 

Very happy to hear, as this is in line with one old idea of mine: https://github.com/corona-warn-app/cwa-app-android/pull/864#issuecomment-674185410 :tada:

Reply:
@d4rken 

> @d4rken proposed creating `googleAPIVersion.requireV15()` and throwing a specific exception at this point,

Anyway, related to my information above, it should be googleAPIVersion.require**V16**(), as you can't query V1.5 as `getVersion()` was introduced in ENF v1.6.

Reply:
https://github.com/corona-warn-app/cwa-app-android/pull/1680